vernac/debugHook.mli

```diff
@@ -39,8 +42,25 @@ module Intf : sig
     (** request a debugger command from the client *)
     ; submit_answer : Answer.t -> unit
     (** receive a debugger answer from Ltac *)
+    ; isCoqtop : bool
+    (** whether the debugger is running in coqtop/coqc *)
```

Comment:
This also seems in the wrong place.

Reply:
@jfehrle , if you needed to add those due to ltac_debug being in a plugin, IMHO that is not going to scale. I'd suggest instead doing as we do in SerAPI and having you code as a plugin that requires the ltac plugin to be loaded, that way you can link to ltac_debug directly. That sounds scary but it is trivial to setup nowadays with the dune build.

[By the way I dislike our plugin system, for resasons like this one, which lead people to have to add hacks, but that's another story]

Reply:
`isCoqtop` indicates whether the debugger is being run from a terminal-only environment, such as `coqtop`, versus a visual environment.  This has two small effects:
- If in a visual environment, don't print the next statement to be executed since that statement will be highlighted in an editor buffer
- Entering the debugger in a terminal-only environment stops at first tactic, essentially a "step-in".  In the visual environment, the debugger doesn't stop at entry--it only stops when it encounters a breakpoint (like a "continue").  I may add a feature to let the user choose which behavior to use.  Note that we may enter the debugger multiple times from the primary script.  In `foo; bar. auto. foo; bar.` we enter the debugger on both of the `foo; bar.` sentences.

When you say "that is not going to scale", I suppose you mean that the code won't be maintainable/understandable?  If so, I don't see why you would believe that.  I'm open to looking at what it would take to convert it to a plugin if you can point me to similar code such as in SerAPI.  But it does sound more complex and less maintainable.  It doesn't help that (AFAIK) some of the plugin details are obscure.

Also I don't see how this would work for `idetop.ml`, unless you're suggesting that we somehow make that part of a plugin.

Is it really worth the effort to convert this to a plugin just to move 2 or 3 variable declarations?  I don't see the benefit.

Reply:
There are two things here:
- how to indeed keep these variables accessible: I think for now they can stay, however having such kind of mutable state amounts to adding technical debt, so while 2 or 3 declarations isolated are not a lot, this is how cruft does accumulate over time and cleanup effort is not free; but I don't think this indeed a big deal for the PR now (meaning: not a blocker, keep them there for now)
- regarding the `isCoqTop`, it seems to me against the spirit of any protocol that it has to keep an state for such kind of setups, instead the protocol should be refined so this is not needed. Otherwise, imagine I have 3rd client (let's say jsCoq), so now, we need to update the protocol with an `isJsCoq` variable? We should not; so this issue is more serious IMO.

Reply:
The code doesn't mutate `isCoqtop`.  AFAIK records don't support immutable fields.  It would be nice if OCaml supported that.  I could convert this to another callback if you like, but the field for the callback would still be mutable.  And I could change the name to something more neutral such as `isVisual`.  If we add a third client that needs different/additional configuration items, we can easily add to or modify the interface at that time.

Your notion of "technical debt" and "cruft" is very expansive.  It seems extreme to avoid any use of records because their fields are mutable.  Note that the adjacent fields `read_cmd` and `submit_answer` are equally mutable.

Reply:
The problem with isCoqtop is not about mutability, my two points are fully independent. IMO the protocol should not have this flag, it is just used to uncover a likely design deficit on the protocol side, and it doesn't scale, neither in testing nor in client setup.

Reply:
> IMO the protocol should not have this flag

I assume you mean "the interface" rather than "the protocol".  What would you have me do?  Add a Db_config message (or messages) to the XML protocol to configure the two behaviors I bullet-pointed above?  Independently configurable?  Or would you prefer that the interface includes a function `is_visual` to pass the setting when called?

> it is just used to uncover a likely design deficit on the protocol side

What is the design deficit?

> it doesn't scale, neither in testing nor in client setup

How is this not scalable?  I don't follow you.

Reply:
The flag should not be needed, as the protocol should provide the data in a form that clients can do all their rendering decisions client-side.

The flag doesn't scale as we cannot add a boolean flag each time a client has a specific display necessity.

Reply:
Each proposed flag should be evaluated on its merits.  It's not reasonable to say we can never have any boolean flags.  At the moment, I have no reason to suggest other flags.  

Let's come back to this point a bit later.  Please look at the rest of the code--let's try to identify all the things that need attention, then we can decide what things need to be addressed pre-merge versus those that can addressed shortly thereafter.  And make a list.